### PR TITLE
Initial pass at importing ShapeShift Trades.

### LINKED ISSUE #3416: Support for import of ShapeShift Trades export csv format
## Abstract

Adding support for new file imports.  I have identified ShapeShift trades are not supported in the import component, but the past release included support for the ShapeShift airdrop. 

## Motivation

Adding support for ShapeShift's ecosystem and gateways such as WYRE and BANXA, which the platform has integrated, will potentially broaden the userbase.

## Specification

Update English locale
Add source `'shapeshift-trades'`
Add a new Location to the schema in sequence for ShapeShift
Update rest API controller
Write import consumer and import_shapeshift_trades_csv
Add Location to `EXTERNAL_EXCHANGES`
Update the import component with a user interface in line with other import sources

### Considerations

The status column can be `complete` or `returned`.  There may be other edge cases to the service which I have not encountered.  I can investigate on the ShapeShift discord channel.

### PR DESCRIPTION
**Pull request by arosboro**
Closes #3416 

## Checklist

- [x] The PR modified the frontend and updated the [user guide](https://github.com/rotki/rotki/blob/develop/docs/usage_guide.rst) to reflect the changes.
- [x] Test coverage
- [ ] Find a spec for the import format
- [x] Edge case for returned trades (should be ignored or does it represent a loss in fees/gas)
- [x] Edge case for symbol DAI to SAI prior to 2019-11-18.
- [x] Descriptive note (in tx, out tx refund tx)
- [x] Invert assets on light mode vs dark mode

### CODE REVIEW

File: frontend/app/src/components/history/consts.ts [hunk 1]
```diff
@@ -181,6 +182,13 @@ export const tradeLocations: TradeLocationData[] = [
     imageIcon: true,
     exchange: true
   },
+  {
+    identifier: EXCHANGE_SHAPESHIFT,
+    name: 'ShapeShift',
+    icon: require('@/assets/images/shapeshift-fox.svg'),
```
Comment (LefterisJP) [inline frontend/app/src/components/history/consts.ts:188]: Is this different than the shapeshit image we have for the fox token?

File: frontend/app/src/views/ImportData.vue [hunk 2]
```diff
@@ -178,6 +178,65 @@
             </v-col>
           </v-row>
           <v-divider class="mt-2 mb-4" />
+          <v-row>
+            <v-col cols="12">
+              <div class="import-data__crypto-com">
+                <v-img
+                  max-width="200"
+                  :src="require('@/assets/images/import/shapeshift.svg')"
```
Comment (LefterisJP) [inline frontend/app/src/views/ImportData.vue:186]: Why not use the same image as "shapeshift-fox.svg"?

File: rotkehlchen/data/importer.py [hunk 3]
```diff
@@ -934,3 +934,67 @@ def import_nexo_csv(self, filepath: Path) -> Tuple[bool, str]:
...
+        trade = Trade(
+            timestamp=timestamp,
+            location=Location.SHAPESHIFT,
+            base_asset=buy_asset,
+            quote_asset=sold_asset,
+            trade_type=TradeType.BUY,
+            amount=buy_amount,
+            rate=Price(rate),
+            fee=Fee(fee),
+            fee_currency=buy_asset,  # Assumption that mineFee is denominated in outputCurrency
```
Comment (LefterisJP) [inline rotkehlchen/data/importer.py:971]: I assume typo

```suggestion
            fee_currency=buy_asset,  # Assumption that minFee is denominated in outputCurrency
```

File: rotkehlchen/tests/api/test_data_import.py [hunk 4]
```diff
@@ -160,6 +161,26 @@ def test_data_import_nexo(rotkehlchen_api_server):
     assert_nexo_results(rotki)
 
 
+@pytest.mark.parametrize('number_of_eth_accounts', [0])
+def test_data_import_shapeshift_trades(rotkehlchen_api_server):
+    """Test that the data import endpoint works successfully for blockfi trades"""
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_data_import.py:166]:
```suggestion
    """Test that the data import endpoint works successfully for shapeshift trades"""
```

**Comment by arosboro** [comment]
I determined that `returned` status shows as completed transactions for the `inputTxid` matching the `inputAmount` and `inputCurrency` and the `refundTxid` for the `outputAmount` and `outputCurrency` in my blockchain accounts verified on ShapeShift.com.  The behaviour is the same as with a `complete` status.

If the `inputAddress` and `outputAddress` are `''`, the change occurred on a DEX through the mobile app using the Native wallet or later when the web application supported DEX trading.  IMO these should be handled by the SushiSwap and Uniswap modules, so the importer ignores them and logs a debug message.

**Review by LefterisJP — COMMENTED** [review]
Some initial comments from a very first quick look. Let us know when the PR is ready for full review.

File: frontend/app/src/components/history/consts.ts (hunk 1 above)
Comment (arosboro) [reply frontend/app/src/components/history/consts.ts:188]: It's pulled from https://drive.google.com/file/d/1NgnWwq78OY323FwxoYQbZj4BQZqVLk-L/view and I just verified it is identical to `frontend/app/src/assets/images/airdrops/FOX_Icon_dark.svg`

File: frontend/app/src/views/ImportData.vue (hunk 2 above)
Comment (arosboro) [reply frontend/app/src/views/ImportData.vue:186]: Keeping in line with the UI/UX of the Import Data component.  Other importers use horizontal branding, and shapeshift provides official horizontal branding at https://shapeshift.com/brand-assets.

File: rotkehlchen/data/importer.py (hunk 3 above)
Comment (arosboro) [reply rotkehlchen/data/importer.py:971]: it is actually `minerFee`.  Wanted to discuss this.  It is not technically an exchange fee, and as such, does it belong reported with the trade as a Fee?  `outputAmount = inputAmount * rate - minerFee`

File: frontend/app/src/components/history/consts.ts (hunk 1 above)
Comment (arosboro) [reply frontend/app/src/components/history/consts.ts]: I reverted to the FOX_Icon_dark.svg in this instance.

**Comment by arosboro** [comment]
This commit marks completion on my end, as long as CI reviews pass it can be a candidate for final review.

File: frontend/app/src/components/history/type.d.ts [hunk 5]
```diff
@@ -1,9 +1,11 @@
+import {VueConstructor} from "vue";
 import { TradeLocation } from '@/services/history/types';
 
 export type TradeLocationData = {
   readonly identifier: TradeLocation;
   readonly name: string;
   readonly icon: string;
   readonly imageIcon: boolean;
+  readonly component: VueConstructor | boolean;
```
Comment (kelsos) [inline frontend/app/src/components/history/type.d.ts:9]: Out of curiosity Is there any specific reason why the `| boolean` approach was selected instead of making it optional and leaving the value undefined if there is no component?

Comment (arosboro) [reply frontend/app/src/components/history/type.d.ts:9]: @kelsos I wasn't sure of the proper syntax.  I will change it to read `readonly component?: VueConstructor;`

File: frontend/app/src/views/ImportData.vue (hunk 2 above)
Comment (arosboro) [reply frontend/app/src/views/ImportData.vue]: I removed the svg assets in favor of components containing the svg paths.

File: rotkehlchen/data/importer.py [hunk 6]
```diff
@@ -934,3 +934,87 @@ def import_nexo_csv(self, filepath: Path) -> Tuple[bool, str]:
...
+        fee = deserialize_fee(csv_row['minerFee'])
+        in_addr = csv_row['inputAddress']
+        out_addr = csv_row['outputAddress']
+        # TODO: determine exact time ShapeShift started using SAI with trades in csv exports
+        sai_sunset_timestamp = deserialize_timestamp_from_date(
+            date='2019-11-18T00:00:00Z',
+            formatstr='iso8601',
+            location='ShapeShift',
+        )
+        notes = f'''
```
Comment (LefterisJP) [inline rotkehlchen/data/importer.py:961]: nitpick: Here and in some other places in the code I see you use `'''`. For the three quotes we actually use `"""` in the codebase.

File: rotkehlchen/data/importer.py [hunk 7]
```diff
@@ -934,3 +934,87 @@ def import_nexo_csv(self, filepath: Path) -> Tuple[bool, str]:
...
+        )
+        buy_asset = symbol_to_asset_or_token(csv_row['outputCurrency'])
+        buy_amount = deserialize_asset_amount(csv_row['outputAmount'])
+        sold_asset = symbol_to_asset_or_token(csv_row['inputCurrency'])
+        sold_amount = deserialize_asset_amount(csv_row['inputAmount'])
+        rate = deserialize_asset_amount(csv_row['rate'])
+        fee = deserialize_fee(csv_row['minerFee'])
+        in_addr = csv_row['inputAddress']
+        out_addr = csv_row['outputAddress']
+        # TODO: determine exact time ShapeShift started using SAI with trades in csv exports
```
Comment (LefterisJP) [inline rotkehlchen/data/importer.py:955]: Is this TODO still needed? you have a timestamp here. If this is indeed the right one, can you move it to the module level? There is no need to recompute it every time `_consume_shapeshift_trade()` is called.

File: rotkehlchen/data/importer.py [hunk 8]
```diff
@@ -934,3 +934,87 @@ def import_nexo_csv(self, filepath: Path) -> Tuple[bool, str]:
...
+'''
+        if sold_amount == ZERO:
+            log.debug(f'Ignoring ShapeShift trade with sold_amount equal to zero. {csv_row}')
+            return
+        if in_addr == '' or out_addr == '':
+            log.debug(f'Ignoring ShapeShift trade which was performed on DEX. {csv_row}')
+            return
+        # Assuming that as DAI contract tx date only SAI existed.
+        # Converting DAI to SAI in buy_asset and sell_asset.
+        if buy_asset == symbol_to_asset_or_token('DAI') and timestamp <= sai_sunset_timestamp:
```
Comment (LefterisJP) [inline rotkehlchen/data/importer.py:978]: no need to use `symbol_to_asset_or_token()` here. This will end up searching the global database for something with a symbol "DAI". A user may also add another "DAI" symbol token and then we will have problems. (in that case the function raises.

What you want is to import `A_SAI` and `A_DAI` from constats/assets.py
https://github.com/rotki/rotki/blob/c4be84e692aa711a1cc7272d33d3209622ec40a0/rotkehlchen/constants/assets.py#L196

**Review by LefterisJP — COMMENTED** [review]
Hey @arosboro thank you so much for the PR!

I left you some comments for small things to fix.

Can you also add a changelog entry?

https://github.com/rotki/rotki/blob/c4be84e692aa711a1cc7272d33d3209622ec40a0/docs/changelog.rst

for issue 3416 https://github.com/rotki/rotki/issues/3416  

Finally the last 3 commits all seem to be for frontend linting so would be nice to have them squashed together.

File: rotkehlchen/data/importer.py (hunk 7 above)
Comment (arosboro) [reply rotkehlchen/data/importer.py:955]: This is the date the MKR DAO launched MCD, so anything before that must have been SAI.  It is possible ShapeShift platform still generated CSVs with DAI until their systems were updated to reflect SAI sometime near this date.

I will move this to the module level and remove the TODO.

Comment (LefterisJP) [reply rotkehlchen/data/importer.py:955]: You should also add it as a timestamp directly. No need to run `deserialize_timestamp_from_date()` since we know the timestamp and it won't change.

Comment (arosboro) [reply rotkehlchen/data/importer.py:955]: @LefterisJP If I understand correctly, this would belong at `rotkehlchen/data/importer.py:1002` before the for loop and should be passed as a parameter to _consume_shapeshift_trade with the Timestamp value.

Is that correct?

Comment (LefterisJP) [reply rotkehlchen/data/importer.py:955]: no no no. You should just add it at the top of the module:

```python
SAI_TIMESTAMP_OR_WHATEVER_NAME_THE_VAR_WILL_HAVE = 1574035200
```

somewhere here: https://github.com/rotki/rotki/blob/c4be84e692aa711a1cc7272d33d3209622ec40a0/rotkehlchen/data/importer.py#L33

**Review by LefterisJP — APPROVED** [review]
Thanks @arosboro. Looks good to me!

**Comment by LefterisJP** [comment]
Failed tests are all due to etherscan 503.
